gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -336,12 +341,50 @@ private ParallelRunner getParallelRunner(FileSystem fs) {
 
   @Override
   public void publishMetadata(Collection<? extends WorkUnitState> states) throws IOException {
-    // Nothing to do
+        for (WorkUnitState workUnitState : states) {
+        publishMetadata(workUnitState);
+    }
   }
 
   @Override
   public void publishMetadata(WorkUnitState state) throws IOException {
-    // Nothing to do
-  }
+    String metadataValue = state.getProp(ConfigurationKeys.DATA_PUBLISHER_METADATA_STR);
+    if (metadataValue == null) {
+      //Nothing to write
+      return;
+    }
+
+    if (state.getProp(ConfigurationKeys.DATA_PUBLISHER_METADATA_OUTPUT_DIR) == null) {
+      LOG.error("Missing metadata output directory path : "  + ConfigurationKeys.DATA_PUBLISHER_METADATA_OUTPUT_DIR 
+                + " in the config");
+      return;
+    }
+
+    //Write for each branch
+    for (int branchId = 0; branchId < this.numBranches; branchId++) {
+      FileSystem fs = this.metaDataWriterFileSystemByBranches.get(branchId);
 
+      String fileName = ForkOperatorUtils.getPropertyNameForBranch("metadata.txt", this.numBranches, branchId);
+      String metaDataOutputDirStr = state.getProp(ConfigurationKeys.DATA_PUBLISHER_METADATA_OUTPUT_DIR);
+
+      Path metaDataOutputPath = new Path(metaDataOutputDirStr);
+      try {
+        if (!fs.exists(metaDataOutputPath)) {
+          WriterUtils.mkdirsWithRecursivePermission(fs, metaDataOutputPath, this.permissions.get(branchId));
+        }
+
+        Path metaFilepath = new Path(metaDataOutputDirStr, fileName);
+
+        //Delete the file if metadata already exists
+        if (fs.exists(metaFilepath)) {
+          HadoopUtils.deletePath(fs, metaFilepath, false);
+        }
+
+        FSDataOutputStream outputStream = this.closer.register(fs.create(metaFilepath));
+        outputStream.write(metadataValue.getBytes());
+      } catch (IOException e) {
+        LOG.error("metadata file is not generated", e);
```

Comment:
Nit: It is a useful pattern to do: LOG.error("Some error message:" + e, e)

This will log the "e" message on the same line and it makes it easy to grep for errors.